### PR TITLE
[Asio] Introduce Async I/O component

### PR DESCRIPTION
**Pull request by azjezz**

### CODE REVIEW

File: src/Psl/Asio/Exception/TimeoutException.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+declare(strict_types=1);
+
+namespace Psl\Asio\Exception;
+
+use Psl\Exception;
+
+final class TimeoutException extends Exception\RuntimeException implements ExceptionInterface
+{
+    public function __construct(string $message = 'operation timed out.', int $code = 0)
```
Comment (BackEndTea) [inline src/Psl/Asio/Exception/TimeoutException.php:9]: Should the exception expose how long it took?

Comment (azjezz) [reply src/Psl/Asio/Exception/TimeoutException.php:9]: probably, will see ( this PR is not getting merged until PHP 8.1 is released, so there's still time to change the API 😛 )

File: src/Psl/Asio/stub.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+declare(strict_types=1);
+
+if (false) {
+    final class Fiber
+    {
+        public function __construct(callable $c)
+        {
+        }
```
Comment (azjezz) [inline src/Psl/Asio/stub.php]: TODO: remove stub before merge

File: src/Psl/Asio/Awaitable.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+     *  (callable(?Throwable, ?T): Awaitable<mixed>)|
+     *  (callable(?Throwable): void)|
+     *  (callable(?Throwable): Awaitable<void>)|
+     *  (callable(?Throwable, ?T): void)|
+     *  (callable(?Throwable, ?T): Awaitable<void>)
+     * ) $callable
+     *
+     * @internal
+     */
+    public function onJoin(callable $callable): void;
```
Comment (veewee) [inline src/Psl/Asio/Awaitable.php:64]: Where does this method name come from?
In amp, they use on resolve, which is more clear IMO.

Comment (azjezz) [reply src/Psl/Asio/Awaitable.php:64]: This method ( maybe the whole class? ) is going to be removed soon thanks to some clever work by @:trowski and @:kelunik.

Comment (veewee) [reply src/Psl/Asio/Awaitable.php:64]: Cool! Is it a loop / promise rfc to php-src perhaps? ;)

**Pull request closed without merging**